network/router.go

```diff
@@ -118,6 +118,17 @@ func (r *Router) Start() {
 			}
 			return
 		}
+
+		// Reject incoming connections from invalid peers
+		if !isPeerValid(dst.ID) {
+			log.Errorf("rejecting incoming connection from %v: invalid peer %v",
```

Comment:
That sounds more like a `log.Warnf` to me - let's take @tharvik 's vote to decide ;)

Reply:
indeed, more like a warning.

Reply:
OK, for my future reference, what is the criteria? :)

Reply:
IMO, it's an expected behavior, a interesting-to-see failure but not one that would normally be acted upon, compared to an "error" that should be investigated.

Reply:
If everything is working as it should, it is not really an expected behavior (maybe indicative of an attack).
But I agree it is a fine line... I will bow to the majority ;-) .

Reply:
Writing this down makes me think there are a lot of errors that should be warnings ;) I had to add `panic` to the list to think straight:

- `Panic` - if the program continues, it might make things worse: db-failure, out-of-memory error, consistency-checks failure that cannot be recovered from, ...
- `Error` - an action **initiated by the system** couldn't be executed, but it's not critical. Probably mostly network failures: couldn't contact a node, invalid reply from the network
- `Warning` - something noteworthy arrived at the system that might be interesting to check out: invalid (connection) request, failed transaction

Like always - once you actually start thinking about it, it's not that easy ;)

For me the difference between error and warning is the initiator: an error is something that the system tried to do and failed, while a warning is a failure while trying to execute an action from the outside.

Which would clean up a lot the failing transactions that are currently `Error`s, but according to this list, they would be `Warning`s. Which make them less: `DO SOMETHING ABOUT IT - NOW!!`

Reply:
I agree with the `Panic`, which pertains to situations from which the software cannot recover.
For me. however, the difference between `Error` and `Warning` is not so much related to the initiator, but to the severity of what happened, a warning indicating an unexpected situation that might or might not lead to something more serious in the future, and an error being a situation already serious.

But as always, the most important thing is to have clear definitions to ensure objectivity in the choices. :)

Examples of common logging libraries that I am familiar with and how they define their levels:
* Python standard `logging` library: https://docs.python.org/3.8/howto/logging.html#logging-basic-tutorial
* `Log4j`, a popular logging library for java: https://en.wikipedia.org/wiki/Log4j#Log4j_log_levels